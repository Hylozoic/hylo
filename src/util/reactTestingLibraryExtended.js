import React from 'react'
import { MemoryRouter } from 'react-router'
import { Provider } from 'react-redux'
import { createStore } from 'redux'
import rootReducer from 'store/reducers'
import createMiddleware from 'store/middleware'

export function generateStore (history, initialState) {
  return createStore(rootReducer, initialState, createMiddleware(history))
}

export const AllTheProviders = (store) => ({ children }) => {
  return (
    <Provider store={store || createStore(() => {})}>
      <MemoryRouter>
        {children}
      </MemoryRouter>
    </Provider>
  )
}

// re-export everything

const rtl = require('react-testing-library')
const customRender = (ui, options, providersFunc) =>
  rtl.render(ui, { wrapper: providersFunc, ...options })

module.exports = {
  ...rtl,
  render: customRender
}
