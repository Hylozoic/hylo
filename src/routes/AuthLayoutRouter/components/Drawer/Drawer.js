import React from 'react'
import { Link, useLocation } from 'react-router-dom'
import { push } from 'connected-react-router'
import { get } from 'lodash/fp'
import { useDispatch, useSelector } from 'react-redux'
import { bgImageStyle } from 'util/index'
import { contextSwitchingUrl, createGroupUrl, groupUrl } from 'util/navigation'
import {
  ALL_GROUPS_ID, ALL_GROUPS_AVATAR_PATH, DEFAULT_AVATAR,
  PUBLIC_CONTEXT_ID, PUBLIC_CONTEXT_AVATAR_PATH
} from 'store/models/Group'
import { toggleDrawer as toggleDrawerAction } from 'routes/AuthLayoutRouter/AuthLayoutRouter.store'
import getMyGroups from 'store/selectors/getMyGroups'
import getCanModerate from 'store/selectors/getCanModerate'
import Badge from 'components/Badge'
import Button from 'components/Button'
import Icon from 'components/Icon'
import cx from 'classnames'
import s from './Drawer.scss' // eslint-disable-line no-unused-vars
import getMe from 'store/selectors/getMe'
import getMyMemberships from 'store/selectors/getMyMemberships'
import { getMemberships } from 'routes/MemberProfile/MemberProfile.store'

export const defaultContexts = [
  {
    id: PUBLIC_CONTEXT_ID,
    name: 'Public Groups & Posts',
    groups: [],
    context: 'public',
    avatarUrl: PUBLIC_CONTEXT_AVATAR_PATH
  },
  {
    id: ALL_GROUPS_ID,
    name: 'All My Groups',
    groups: [],
    context: 'all',
    avatarUrl: ALL_GROUPS_AVATAR_PATH
  }
]

export default function Drawer (props) {
  const dispatch = useDispatch()
  const groups = useSelector(getMyGroups)
  const canModerate = useSelector(state => props.group && getCanModerate(state, props))
  const currentLocation = useLocation()
  const routeParams = props.match.params
  const { group, className } = props

  const toggleDrawer = () => dispatch(toggleDrawerAction())
  const goToCreateGroup = () => dispatch(push(createGroupUrl(get('match.params', props))))

  let bannerUrl

  if (!group) {
    bannerUrl = ''
  } else {
    ({ bannerUrl } = group)
  }

  return (
    <div className={className} styleName='s.groupDrawer'>
      <div styleName={cx('drawerHeader', { 's.currentGroup': group !== null })} style={bgImageStyle(bannerUrl)}>
        <div styleName='drawerBanner'>
          <div styleName='s.hyloLogoBar'>
            <img src='/hylo.svg' width='50px' />
            <Icon name='Ex' styleName='s.closeDrawer' onClick={toggleDrawer} />
          </div>
          <Logo group={group} />
          {canModerate && (
            <Link styleName='s.settingsLink' to={groupUrl(group.slug, 'settings')}>
              <Icon name='Settings' styleName='s.settingsIcon' /> Group Settings
            </Link>
          )}
        </div>
        <div styleName='backgroundFade' />
      </div>
      <div>
        <ul styleName='s.groupsList'>
          {defaultContexts && defaultContexts.map(context =>
            <ContextRow currentLocation={currentLocation} group={context} routeParams={routeParams} key={context.id} />
          )}
        </ul>
        <ul styleName='s.groupsList'>
          <li styleName={cx('s.sectionTitle', 's.sectionTitleSeparator')}>My Groups</li>
          {groups.map(group =>
            <ContextRow currentLocation={currentLocation} group={group} routeParams={routeParams} key={group.id} />
          )}
        </ul>
        <div styleName='s.newGroup'>
          <Button
            color='white'
            styleName='s.newGroupBtn'
            label='Start a Group'
            onClick={goToCreateGroup}
          />
        </div>
      </div>
    </div>
  )
}

export function ContextRow ({ currentLocation, group, routeParams }) {
  const { avatarUrl, context, name, newPostCount, slug } = group
  const imageStyle = bgImageStyle(avatarUrl || DEFAULT_AVATAR)
  const showBadge = newPostCount > 0
  const path = contextSwitchingUrl({ groupSlug: slug, context }, routeParams)
  return (
    <li styleName={cx('s.contextRow', { 's.currentContext': currentLocation === path })}>
      <Link to={contextSwitchingUrl({ context: context || 'groups', groupSlug: slug }, routeParams)} styleName='s.contextRowLink' title={name}>
        <div styleName='s.contextRowAvatar' style={imageStyle} />
        <span styleName='s.group-name'>{name}</span>
        {showBadge && <Badge expanded number={newPostCount} />}
      </Link>
    </li>
  )
}

function Logo ({ group }) {
  if (!group) return null
  const { slug, name, location, avatarUrl } = group
  return (
    <Link styleName='s.currentGroup' to={groupUrl(slug)}>
      <div styleName='s.avatar' style={bgImageStyle(avatarUrl || DEFAULT_AVATAR)} />
      <div className='drawer-inv-bd'>{name}</div>
      <div className='drawer-inv-sm'>{location}</div>
    </Link>
  )
}
