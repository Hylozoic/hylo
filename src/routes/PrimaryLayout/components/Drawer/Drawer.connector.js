import { push } from 'connected-react-router'
import { get } from 'lodash/fp'
import { connect } from 'react-redux'
import { createSelector } from 'reselect'

import { toggleDrawer } from 'routes/PrimaryLayout/PrimaryLayout.store'
import { ALL_GROUPS_ID, ALL_GROUPS_AVATAR_PATH, PUBLIC_CONTEXT_ID, PUBLIC_CONTEXT_AVATAR_PATH } from 'store/models/Group'
import getMyMemberships from 'store/selectors/getMyMemberships'
import getMe from 'store/selectors/getMe'
import { createGroupUrl } from 'util/navigation'

const defaultContexts = [
  {
    id: PUBLIC_CONTEXT_ID,
    name: 'Public Groups & Posts',
    groups: [],
    context: 'public',
    avatarUrl: PUBLIC_CONTEXT_AVATAR_PATH
  },
  {
    id: ALL_GROUPS_ID,
    name: 'All My Groups',
    groups: [],
    context: 'all',
    avatarUrl: ALL_GROUPS_AVATAR_PATH
  }
]

const getGroups = createSelector(
  getMyMemberships,
  (memberships) => memberships.map(m => ({ ...m.group.ref, newPostCount: m.newPostCount })).sort((a, b) => a.name.localeCompare(b.name))
)

export function mapStateToProps (state, props) {
  const { currentLocation } = state.locationHistory
  const groups = getGroups(state)
  const canModerate = props.group && getMe(state, props).canModerate(props.group)

  return {
    currentLocation,
    defaultContexts,
    groups,
    canModerate
  }
}

export function mapDispatchToProps (dispatch, props) {
  return {
    toggleDrawer: () => dispatch(toggleDrawer()),
    goToCreateGroup: () => dispatch(push(createGroupUrl(get('match.params', props))))
  }
}

export default connect(mapStateToProps, mapDispatchToProps)
