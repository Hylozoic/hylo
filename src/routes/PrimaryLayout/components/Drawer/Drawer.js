import PropTypes from 'prop-types'
import React from 'react'
import { Link } from 'react-router-dom'
import { bgImageStyle } from 'util/index'
import { contextSwitchingUrl, groupUrl } from 'util/navigation'
import Badge from 'components/Badge'
import Button from 'components/Button'
import Icon from 'components/Icon'
import s from './Drawer.scss' // eslint-disable-line no-unused-vars
import { DEFAULT_AVATAR } from 'store/models/Group'
import cx from 'classnames'

const { string, number, arrayOf, shape } = PropTypes

export default class Drawer extends React.PureComponent {
  static propTypes = {
    groups: arrayOf(shape({
      id: string,
      newPostCount: number,
      name: string,
      slug: string,
      avatarUrl: string
    }))
  }

  render () {
    const { currentLocation, group, groups, defaultContexts, className, toggleDrawer, canModerate } = this.props
    const routeParams = this.props.match.params

    let bannerUrl

    if (!group) {
      bannerUrl = ''
    } else {
      ({ bannerUrl } = group)
    }

    return <div className={className} styleName='s.groupDrawer'>
      <div styleName={cx('drawerHeader', { 's.currentGroup': group !== null })} style={bgImageStyle(bannerUrl)}>
        <div styleName='drawerBanner'>
          <div styleName='s.hyloLogoBar'>
            <img src='/hylo.svg' width='50px' />
            <Icon name='Ex' styleName='s.closeDrawer' onClick={toggleDrawer} />
          </div>
          <Logo group={group} />
          {canModerate && <Link styleName='s.settingsLink' to={groupUrl(group.slug, 'settings')}>
            <Icon name='Settings' styleName='s.settingsIcon' /> Group Settings
          </Link>}
        </div>
        <div styleName='backgroundFade' />
      </div>
      <div>
        <ul styleName='s.groupsList'>
          {defaultContexts && defaultContexts.map(context =>
            <ContextRow currentLocation={currentLocation} group={context} routeParams={routeParams} key={context.id} />
          )}
        </ul>
        <ul styleName='s.groupsList'>
          <li styleName={cx('s.sectionTitle', 's.sectionTitleSeparator')}>My Groups</li>
          {groups.map(group =>
            <ContextRow currentLocation={currentLocation} group={group} routeParams={routeParams} key={group.id} />
          )}
        </ul>
        <div styleName='s.newGroup'>
          <Button
            color='white'
            styleName='s.newGroupBtn'
            label='Start a Group'
            onClick={this.props.goToCreateGroup}
          />
        </div>
      </div>
    </div>
  }
}

export function ContextRow ({ currentLocation, group, routeParams }) {
  const { avatarUrl, context, name, newPostCount, slug } = group
  const imageStyle = bgImageStyle(avatarUrl || DEFAULT_AVATAR)
  const showBadge = newPostCount > 0
  const path = contextSwitchingUrl({ groupSlug: slug, context }, routeParams)
  return <li styleName={cx('s.contextRow', { 's.currentContext': currentLocation === path })}>
    <Link to={contextSwitchingUrl({ context: context || 'groups', groupSlug: slug }, routeParams)} styleName='s.contextRowLink' title={name}>
      <div styleName='s.contextRowAvatar' style={imageStyle} />
      <span styleName='s.group-name'>{name}</span>
      {showBadge && <Badge expanded number={newPostCount} />}
    </Link>
  </li>
}

function Logo ({ group }) {
  if (!group) return null
  const { slug, name, location, avatarUrl } = group
  return <Link styleName='s.currentGroup' to={groupUrl(slug)}>
    <div styleName='s.avatar' style={bgImageStyle(avatarUrl || DEFAULT_AVATAR)} />
    <div className='drawer-inv-bd'>{name}</div>
    <div className='drawer-inv-sm'>{location}</div>
  </Link>
}
