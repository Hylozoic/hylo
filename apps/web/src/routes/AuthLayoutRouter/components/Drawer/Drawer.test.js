import React from 'react'
import orm from 'store/models'
import extractModelsForTest from 'util/testing/extractModelsForTest'
import { AllTheProviders, render, screen } from 'util/testing/reactTestingLibraryExtended'
import Drawer, { ContextRow } from './Drawer'

const fooGroup = {
  id: '11',
  slug: 'foo',
  name: 'Foomunity',
  avatarUrl: '/foo.png'
}

const barGroup = {
  id: '22',
  slug: 'bar',
  name: 'Barmunity',
  avatarUrl: '/bar.png'
}

function currentUserWithGroupsProvider () {
  const ormSession = orm.mutableSession(orm.getEmptyState())
  const reduxState = { orm: ormSession.state }

  extractModelsForTest({
    me: {
      id: '1',
      name: 'Test User',
      hasRegistered: true,
      emailValidated: true,
      settings: {
        signupInProgress: false
      },
      memberships: [
        {
          id: '2',
          person: {
            id: '1'
          },
          newPostCount: 0,
          group: fooGroup
        },
        {
          id: '3',
          person: {
            id: '1'
          },
          newPostCount: 7,
          group: barGroup
        }
      ]
    }
  }, 'Me', ormSession)

  return AllTheProviders(reduxState)
}

const match = {
  match: {
    params: {
      context: 'groups',
      groupSlug: 'slug'
    }
  }
}

it('shows groups for current user', () => {
  render(
    <Drawer match={match} />,
    { wrapper: currentUserWithGroupsProvider() }
  )

  expect(screen.getByText(fooGroup.name)).toBeInTheDocument()
  expect(screen.getByText(barGroup.name)).toBeInTheDocument()
})

describe('ContextRow', () => {
  it('renders with zero new posts', () => {
    render(<ContextRow group={fooGroup} />)

    expect(screen.getByText(fooGroup.name)).toBeInTheDocument()
    expect(screen.queryByRole('status')).not.toBeInTheDocument() // Assuming Badge is rendered with role="status"
  })

  it('renders with new posts', () => {
    render(<ContextRow group={barGroup} />)

    expect(screen.getByText(barGroup.name)).toBeInTheDocument()
    expect(screen.getByRole('status')).toHaveTextContent('7') // Assuming Badge is rendered with role="status"
  })
})
