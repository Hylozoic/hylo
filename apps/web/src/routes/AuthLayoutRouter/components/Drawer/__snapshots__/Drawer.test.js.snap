// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContextRow renders with new posts 1`] = `
<li
  data-stylename="s.contextRow"
>
  <Link
    data-stylename="s.contextRowLink"
    title="Barmunity"
    to="/groups/bar"
  >
    <div
      data-stylename="s.contextRowAvatar"
      style={
        Object {
          "backgroundImage": "url(/bar.png)",
        }
      }
    />
    <span
      data-stylename="s.group-name"
    >
      Barmunity
    </span>
  </Link>
</li>
`;
